[WIP] Proportional fonts support

thjbdvlt opened the pull request:
This work-in-progress attempt to add support for proportional (i.e. non-monospace) fonts, as discussed in #127.

## Motivation

As a terminal application, Vim/Neovim have no support for non-monospace fonts, while many Graphical editors have: Emacs, VS Code, and many others. This is limiting the use of Neovim for non-programming (prose) writing practices, despite its value for such stuff: academic texts[^1],  novels, poetry [^3]  --whatever writing task that requires not so much to write text but to "rewrite and edit[^2]" it--, or everyday tasks such as mail, personal notes... 
For prose, proportional fonts have strong benefits (easier to read, especially for long texts), but [even for code](https://storck.io/posts/proportional-programming-code-fonts/), some people argue that proportional fonts are worth to use (see the [Go Fonts](https://go.dev/blog/go-fonts)).

The design of Goneovim seems to make possible the use of proportional fonts for Neovim. It would be great, since there is no GUI that support that feature (Neovide, Neovim-Qt, Fvim, Uivonim...).

[^1]: There is a lot of plugins and tools to use (Neo)Vim with [Pandoc](https://github.com/vim-pandoc/vim-pandoc), or [Bib(La)TeX](https://github.com/ferdinandyb/bibtexcite.vim), or [Zotero](https://github.com/jalvesaq/zotcite).

[^2]: Howard Becker, [Writing for Social Scientists](https://atgender.eu/wp-content/uploads/sites/207/2022/03/Howard-S.-Becker-Writing-for-Social-Scientists_-How-to-Start-and-Finish-Your-Thesis-Book-or-Article_-Second-Edition-University-Of-Chicago-Press-2007.pdf).

[^3]: There are many blog posts stating that (Neo)Vim is a good prose editor. See, e.g.:  https://jonathanh.co.uk/blog/writing-prose-in-vim/ or https://discover.hubpages.com/technology/how-and-why-i-use-vim-for-prose-writing or https://jamierubin.net/2019/03/21/writing-with-vim/

## Current Progress

- Text, background and decoration rendering.
- Cursor position, text, background and shape.
- Align the Line Number column end, to align the beginning of the buffer lines.
- Left align buffer content (independantly of gutter/signcolumn content).
- Possibility to use both proportional fonts for some grids and monospace for others.

![victor-burgin](https://github.com/user-attachments/assets/cbb1c3cb-415a-4142-b6a1-b6c54afb8cdd)

<!--
(Left: without the current PR changes / Right: with these changes.)

![2](https://github.com/user-attachments/assets/df468f16-051b-4468-9bc6-a683eadb5707)

![1](https://github.com/user-attachments/assets/da0d3ef7-4e4e-47b1-b087-b6b3d5166ed6)
-->

## Things to be Fixed

- Floating windows borders.
- Floating window position (completion / LSP).
- CPU usage.
- Autocommands such as `autocmd BufEnter *.md :GonvimGridFontAutomaticHeight "Liberation Sans"`.
- Completion window position at the right place (near cursor).
- Some punctuation having incorrect width depending on the context:

![brackets](https://github.com/user-attachments/assets/6dc8e509-34c1-46ed-930f-3d388bf5f959)
![parentheses](https://github.com/user-attachments/assets/b2207abe-3ed0-42cc-ba3b-ff35b04795f3)
![ampersand](https://github.com/user-attachments/assets/75a59f7f-7e40-46c1-8673-6acf4d6eab67)

## Current Limitation: Visual Wrap

The linebreak (visual wrap) is done, as now, by Neovim (using optiong `linebreak`) and not by Goneovim. Thus the linebreak happens *long before* the horizontal end (right) of the window (see screenshot), because it's based on the widest character (italic `W`). There are many options here. Goneovim could manage by itself the wrapping, but I don't know how much work it would require, and how it could impact performance.
On the other hand, as the screenshot shows, when writing prose, line usually tends to have no extreme widths: neither as narrow as lines fulls of `i`, nor as wide as lines full of uppercased italic `W`. Width seems not to vary so much. Thus, the size of the window could be based on the length of a letter narrower than `W`, e.g. `u`. (But I don't know what to do / how to handle special cases, when line is too long.) At the same time, line break long before the horizontal limit of the window may not be an issue at all for some users, as short lines (~10 words) as usually considered easier to read than long lines.

![visual-wrap](https://github.com/user-attachments/assets/0c358b4d-9f2e-4151-b244-769b81e855ed)

## Configuration Requirements

It's currently required to set some configuration variable in order to use proportional fonts:

- `DisableLigatures` must be set to `false`, because if set to `true` it will trigger the cell-based rendering. There's probably no reason to change this.
- `LetterSpace` must be set to `0` for the same reason.

## Changes and Implementation

The changes must not impact performances of Goneovim for non-monospace fonts. Thus, every change is wrapped inside a conditional `if !ont.proportional`. (There may be a small impact because of these tests, but it may not be significant.) These changes use the new `proportional` field, added to the `Font` structure. There are three more fields added to this structure, all of them about `QFontMetrics`: `italicFontMetrics`, `boldFontMetrics`, `italicBoldFontMetrics`. That's because in proportional fonts, each character for each font variant has a specific width. Thus, the existing field `italicWidth` or a new field `italicWidthScale` couldn't be used to compute the width of italic/bold characters. 
The structure `Window` also have a new field, also unused when monospace fonts are used: `xPixelsIndexes`. It's used to store the position, in X-axis, of each character for the lines currently drawn. It's useful to avoid hundreds of calls to `QFontMetrics.HorizontalAdvance(char, ...)`.

thjbdvlt committed:
[WIP] Proportional font: basic support

Implement basic support for proportional font:
- foreground, background, decoration
- proportional cursor width

akiyosi commented:
@thjbdvlt 
Thank you for the PR! I’ll be reviewing the details more thoroughly soon, but I wanted to share some initial thoughts.

Since Neovim’s UI is fundamentally grid-based, simply setting a proportional font as guifont would break the layout of all UI elements. While I don’t intend to restrict users from setting proportional fonts in guifont, I wonder if we could consider an approach where proportional fonts are configured separately using a dedicated command, and applied only to specific grids.

For example, we might imagine a command like GonvimProportionalFontView "some-proportional-font-name", which would allow certain content grids to be rendered using a proportional font, while keeping the base UI rendered with a monospaced font. This concept would be similar to the current GonvimGridFont command, but tailored for proportional font usage.

Additionally, when rendering with proportional fonts, we would likely need to avoid drawing float window borders using the font itself. Instead, we could consider rendering alternative borders using the GUI layer to maintain visual consistency.

Looking forward to discussing this further!

thjbdvlt commented:
@akiyosi Thank you for your thoughts!

Having a special command to set a proportional font for a specific window seems to be the best option. I would typically have an autocmd for markdown files that would set a proportional font for the current window. (So I could keep coding with Inconsolata.)

I think that the modifications proposed in this draft are actually very compatible with this option, as I mostly focused on making the current rendering functions compatible with proportional fonts. It typically works with GonvimGridFont (I just had to add two lines in the function I wrote to compute cell positions; in [this commit](https://github.com/akiyosi/goneovim/pull/591/commits/090e0dbdc92d26e4a42aa23c8b4a712c0324c454)). Another function could be written, but GonvimGridFont could be used as well, without any change, since the `proportional` field in `Font` struct is queried directly to the font (using Qt). I guess that nobody want to use proportional fonts rendered in cells like monospace ones anyway, because it looks terrible.

The only issue  with GonvimGridFont is about resizing, but it seems that it's an issue even with monospace fonts?

I had not yet tried anything to fix floating windows, because with these changes these looks terrible, because of the right border, as you noticed. Completion menu is another issue: it's always incorrectly positioned.

One question I have is: instead of modifying existing functions, like I've done here adding conditions `if font.proportional {...}`, should I make new one, dedicated to proportional fonts? I didn't think it was good, because the changes to those functions are overall small, so it would mostly looks like duplication. But maybe you prefer to separate both rendering logics.

akiyosi commented:
@thjbdvlt 
You're right that using GonvimGridFont for proportional fonts is one possible approach. However, the GonvimGridFont command is more about assigning a specific font to a given grid, and I feel that mixing this with the proportional font use case might not be ideal. (I do recognize that there are still issues with GonvimGridFont, including resizing.)

When a proportional font is specified, maybe the best approach is to apply it to all grids except the global grid (gridid=1) and the message grid (usually gridid=3), which would continue to use the default monospace font. That way, code editing and core UI elements remain stable and predictable, while other content (like markdown previews) can benefit from proportional rendering. That said, as you pointed out, figuring out how to properly position the completion menu remains a major challenge.

Regarding floating windows, goneovim already has logic to draw borders. I think we could build on that mechanism to support proportional fonts, rather than relying on the right-side cell rendering.

As for your last question: if most of the logic is shared, then adding `if font.proportional { ... }` branches within existing functions seems totally fine to me. I don't think it's worth duplicating entire functions just for that case.

thjbdvlt committed:
[WIP] Proportional font: completion windows

For the purpose of completion floating windows, every floating window
that has, in its config, `relative = "cursor"` or `relative = "editor"`
will be positioned relatively to the actual text on the buffer.
It should be a temporary solution until (I hope) neovim implement a way to
identify floating windows that are positioned near the cursor.
(See: https://github.com/neovim/neovim/issues/34595)
A consequence of this commit is that some floating window that must be
centered won't be centered anymore, since their X coordinate will
depend on the text that the buffer contained. But it may not be a real
issue, since the text itself, because of visual wrapping, doesn't occupy
the whole window.
(Note: as other commits, only proportional fonts are affected.)

thjbdvlt commented:
@akiyosi I'll go that way, excluding global and message grids, and using the mechanisms for external windows to draw floating windows.
I've implemented in [this commit](https://github.com/akiyosi/goneovim/pull/591/commits/4d2ae3843dfbbcebf5e4b556c40bc27ae0957d42) a solution for completion (or LSP, or other cursor-based) floating window, which is not perfect (thus, I hope, temporary), because there currently exists no reliable way to check if a floating window must be position relatively to the cursor or not:  `relative = "cursor"` is rarely used plugins, see [this issue](https://github.com/neovim/neovim/issues/34595) I've opened on neovim.

thjbdvlt committed:
[WIP] Proportional font: text offset

Problem:
The Sign Column, containing the Line Numbers and other informations,
has not always the same width, with Proportional Fonts, leading to
misalignement in the text, especially disturbing when using relative
numbers, since the passage from 9 to 10, replacing a space by a digit,
moves the line to the right.

Solution:
Compute the length of the gutter, using `getwininfo` and extract the
field `textoff`. Characters in the gutter all have a fixed-width.
As it can leads to more computation, a global option has been added,
`Editor.ProportionalFontAlignGutter`, that can deactivate this feature.

thjbdvlt committed:
[WIP] Proportional font: GonvimGridFont alternative

Add a new function, similar to GonvimGridFont, but that automatically
calculate the height of the new font, to avoid resizing issues.

thjbdvlt committed:
[WIP] Proportional font: Rebase to master

+ minor changes

akiyosi commented:
@thjbdvlt 
Thanks! I'm currently testing your branch.

It looks like some of the floating window positioning logic is still based on cell width, which doesn't work well with proportional fonts. I think this needs to be adjusted to use pixel-based positioning instead. For example, part of the Telescope window appears overlapped on my screen due to this.

This seems to be related to logic in functions like `setFloatWindowPosition()` in window.go.
<img width="1317" alt="image" src="https://github.com/user-attachments/assets/7cabe286-9b40-4f5d-9511-3521ff2b4f3a" />

thjbdvlt commented:
@akiyosi 
May I ask you which fonts you are using on the screenshot? I think it could be an issue when the proportional font is wider than the global monospace font. (But I don't understand yet.) I can reproduce with Iosevka + Liberation Sans, while it's fine with Iosevka + Linux Libertine O, or with Inconsolata + Liberation Sans.

One issue here is that there are many floating windows, and that the position of the right one depends on the width of the left-one, so we need to use the cellwidth of the floating window and the pixel position from the anchor window.

Also, I have thought about visual-wrap, which is (I think) one of the hardest issue, and I would like to know you opinion about that.

I see two solutions.
1. The first, that doesn't seem reasonable to me, would be to manage the visual-wrap entirely in Goneovim, and to add `:GonvimVisualUp`/`Down` to replace builtins `gk`/`gj`. (I don't think that there is a way to tell to Neovim that *this line is wrapped at 50h character, this one at 74th, this one at 84th...*, so Neovim could manage by itself `gk`/`gj`; but I may be wrong.)
2. Another solution would be to let Neovim in charge of visual wrap. When writing prose, lines tends to have an average width, so the wrap by Neovim actually produces overall good results (see screenshot below). But there could be extreme cases, like a like full in `W`. In these cases, that could be easily identified, we can artificially make the whole line narrower (using the QFont), or smaller, or we can truncate it with a special sign. (I think narrowing would be preferable; it would be similar to print practices with spaces for justification).
2.1. We can also increase the window columns to reduce the unused space on the right.
2.2. It would even be possible (using the same logic as narrowing) to optionally justify the text (using spaces).

![hedgehog](https://github.com/user-attachments/assets/73de858f-3c1c-4167-8d6e-68d73225f4e0)

(Article [Hedgehog](https://en.wikipedia.org/wiki/Hedgehog) on Wikipedia.)

akiyosi commented:
@thjbdvlt 
You're right — after commenting out that section of code, the issue with the Telescope floating window is resolved.

Another issue I noticed is that on lines with many "W" characters, the width of the text cache image is insufficient. The following change is needed to fix this.

``` diff
@@ -3012,9 +3011,14 @@ func (w *Window) newTextCache(text string, hlkey HlKey, isNormalWidth bool) *gui
                )
        }
 
-       width := float64(len(text))*font.cellwidth + 1
-       if hlkey.italic {
-               width = float64(len(text))*font.italicWidth + 1
+       var width float64
+       if !fontfallbacked.proportional {
+               width = float64(len(text))*font.cellwidth + 1
+               if hlkey.italic {
+                       width = float64(len(text))*font.italicWidth + 1
+               }
+       } else {
+               width = fontfallbacked.fontMetrics.HorizontalAdvance(text, -1)
        }

```

thjbdvlt commented:
@akiyosi
I have a  question regarding the function `Window.paint()`.
I've finally manage to fix a bug causing background/decoration to stop before the end of the line when using proportional font.
![decoration-bg](https://github.com/user-attachments/assets/8c8edf89-425b-471a-a9af-f675d9433a43)
The fix requires to change the definitions of `cols` in `Window.paint()`, which is currently defined as:

```go
cols := int(math.Ceil(float64(rect.Width()) / font.cellwidth))
```

And I need to add:

```go
if font.proportional && w.cols > cols {
    cols = w.cols
}
```

I've seen that in the function `window.drawForeground()`, called within `window.paint()`, the `cols` argument is not used anymore to draw text. Instead, `w.cols` is used. so I was wondering why `w.cols` isn't used in the first place? Should I avoid using it for proportional font?

```go
// w.drawText(p, y, col, cols)
w.drawText(p, y, 0, w.cols)
w.drawTextDecoration(p, y, col, cols)
```

thjbdvlt commented:
@akiyosi 
Thank's for the explanation!
(I found another solution, a bit more complex but that avoids drawing the whole line.)

akiyosi commented on editor/window.go:
```diff
@@ -2824,7 +3014,12 @@ func (w *Window) newTextCache(text string, hlkey HlKey, isNormalWidth bool) *gui
 	width := float64(len(text))*font.cellwidth + 1
-	if hlkey.italic {
+	if fontfallbacked.proportional {
+		fmWidth := fontfallbacked.fontMetrics.HorizontalAdvance(text, -1)
```
With proportional fonts, bold styles tend to increase the character width even further, so I think that additional consideration for this case is necessary.

thjbdvlt replied:
You're right. It can be done easily, I think, e.g. using this small function I've added:

```go
func getFontMetrics(font *Font,  highlight *Highlight) *gui.QFontMetricsF {
	if !highlight.italic {
		if !highlight.bold {
			return font.fontMetrics
		}
		return font.boldFontMetrics
	}
	if !highlight.bold {
		return font.italicFontMetrics
	}
	return font.italicBoldFontMetrics
}
```

The following change can thus be enough:

```diff
diff --git a/editor/window.go b/editor/window.go
index 0cbaa3a..71a6342 100644
--- a/editor/window.go
+++ b/editor/window.go
@@ -3015,7 +3015,8 @@ func (w *Window) newTextCache(text string, hlkey HlKey, isNormalWidth bool) *gui
 
 	width := float64(len(text))*font.cellwidth + 1
 	if fontfallbacked.proportional {
-		fmWidth := fontfallbacked.fontMetrics.HorizontalAdvance(text, -1)
+		fm := getFontMetrics(fontfallbacked, &Highlight{bold: hlkey.bold, italic: hlkey.italic})
+		fmWidth := fm.HorizontalAdvance(text, -1)
 		if fmWidth > width {
 			width = fmWidth
 		}
```

(It seems to work, I'll do some tests.)

akiyosi commented:
@thjbdvlt 
I'm planning to apply equivalent changes to `master` as mentioned here: [https://github.com/akiyosi/goneovim/pull/591#issuecomment-3036729310](https://github.com/akiyosi/goneovim/pull/591#issuecomment-3036729310).
Once this PR is ready to be merged, I assume the commit history will be cleaned up and adjusted with appropriate granularity.
So even if there are some temporary inconsistencies with `master`, I think it's fine to resolve them afterward.

thjbdvlt commented:
@akiyosi
I'll rebase to master when the changes will be applied.
I think that, at least for the current state of the PR, 3-4 commits should be enough, maybe even less?
(Most commits are actually fixes of aspects of the initial modifications I haven't thought about.)

Also, I've seen that a function I've added to compute left-column width (`getTextOff`) works well but has significant CPU usage. Do you think I should remove it, or it's OK to let it (I've added it as a non-default option)?

```go
/* Get the buffer offset, i.e. SignColumn and Numbers. */
func (w *Window) getTextOff(delayMs time.Duration) (int, bool) {
	if !editor.config.Editor.ProportionalFontAlignGutter {
		return 0, true
	}
	if w.isFloatWin || w.isMsgGrid || w.isPopupmenu {
		return 0, false
	}
	var isValid bool
	validCh := make(chan bool)
	errCh := make(chan error)
	go func() {
		result, err := w.s.ws.nvim.IsWindowValid(w.id)
		if err != nil {
			errCh <- err
			validCh <- false
			return
		} else {
			validCh <- result
			errCh <- nil
			return
		}
	}()
	select {
	case <- errCh:
		return 0, false
	case isValid = <-validCh:
	case <-time.After(delayMs * time.Millisecond):
		// If the delay is over, returns the last stored version.
		// This avoids flickering.
		return w.endGutterIdx, false
	}
	if !isValid {
		return 0, false
	}
	var output any
	err := w.s.ws.nvim.Call("getwininfo", &output, w.id)
	if err != nil {
		return 0, false
	}
	outputArr, ok := output.([]any)
	if !ok || len(outputArr) < 1 {
		return 0, false
	}
	outputMap, ok := outputArr[0].(map[string]any)
	if !ok {
		return 0, false
	}
	textoff, ok := outputMap["textoff"]
	if !ok {
		return 0, false
	}
	textoffInt, ok := textoff.(int64)
	if !ok {
		return 0, false
	}
	w.endGutterIdx = int(textoffInt)
	return w.endGutterIdx, true
}
```

akiyosi commented:
@thjbdvlt 
Thanks! I also think 3–4 commits are enough.

My understanding is that getTextOff() is there to stabilize the gutter width. I’m in favor of keeping that functionality, but we do need to factor in the performance cost—and I don’t have a good solution for that yet either. The existing isSignColumn() check is incomplete, but moving the decision there (or improving it) could be one possible direction.

thjbdvlt commented:
@akiyosi

I've cleaned the PR history. I'll rebase to current master within a few days.
I'll also try to improve `isExistingColumn` so we can use it for proportional fonts (if I remember well, it doesn't handle yet the spaces in the beginning of wrapped lines, but I need to check).

thjbdvlt committed:
[WIP] Proportional font: Rebase to master

thjbdvlt commented:
@akiyosi I understand that it takes time, there are a lot of changes! And I need to rebase to master again anyway.
